spiffe/tls_peer.go

```diff
@@ -0,0 +1,319 @@
+	conn, err := new(net.Dialer).DialContext(ctx, network, address)
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := p.GetConfig(ctx, expectPeer)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	return tls.Client(conn, config), nil
+}
+
+func (p *TLSPeer) Listen(ctx context.Context, network, address string, expectPeer ExpectPeerFunc) (net.Listener, error) {
+	inner, err := net.Listen(network, address)
+	if err != nil {
+		return nil, err
+	}
+
+	l, err := p.NewListener(ctx, inner, expectPeer)
+	if err != nil {
+		inner.Close()
+		return nil, err
+	}
+
+	return l, nil
+}
+
+func (p *TLSPeer) NewListener(ctx context.Context, inner net.Listener, expectPeer ExpectPeerFunc) (net.Listener, error) {
+	config, err := p.GetConfig(ctx, expectPeer)
+	if err != nil {
+		return nil, err
+	}
+
+	return &peerTLSListener{
+		Listener: inner,
+		config:   config,
+	}, nil
+}
+
+func (p *TLSPeer) GetConfig(ctx context.Context, expectPeer ExpectPeerFunc) (*tls.Config, error) {
+	if expectPeer == nil {
+		return nil, errors.New("authorize callback is required")
+	}
+	if err := p.WaitUntilReady(ctx); err != nil {
+		return nil, err
+	}
+	return &tls.Config{
+		ClientAuth:            tls.RequireAnyClientCert,
+		InsecureSkipVerify:    true,
+		GetCertificate:        AdaptGetCertificate(p),
+		GetClientCertificate:  AdaptGetClientCertificate(p),
+		VerifyPeerCertificate: AdaptVerifyPeerCertificate(p, expectPeer),
+	}, nil
+}
+
+func AdaptGetCertificate(p *TLSPeer) func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+	return func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+		return p.GetCertificate()
+	}
+}
+
+func AdaptGetClientCertificate(p *TLSPeer) func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+	return func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+		return p.GetCertificate()
+	}
+}
+
+func AdaptVerifyPeerCertificate(p *TLSPeer, expectPeer ExpectPeerFunc) func([][]byte, [][]*x509.Certificate) error {
+	return func(rawCerts [][]byte, _ [][]*x509.Certificate) error {
+		var certs []*x509.Certificate
+		for _, rawCert := range rawCerts {
+			cert, err := x509.ParseCertificate(rawCert)
+			if err != nil {
+				return err
+			}
+			certs = append(certs, cert)
+		}
+
+		roots, err := p.GetRoots()
+		if err != nil {
+			return err
+		}
+		if _, err := VerifyPeerCertificate(certs, roots, expectPeer); err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func (p *TLSPeer) DialGRPC(ctx context.Context, addr string, expectPeer ExpectPeerFunc, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
+	config, err := p.GetConfig(ctx, expectPeer)
+	if err != nil {
+		return nil, err
+	}
+	return grpc.DialContext(ctx, addr, append([]grpc.DialOption{
+		grpc.WithTransportCredentials(credentials.NewTLS(config)),
+	}, opts...)...)
+}
+
+func ListenTLS(ctx context.Context, network, addr string, expectPeer ExpectPeerFunc) (net.Listener, error) {
```

Comment:
What's the intended usage/difference between this and Listen? I find it a little confusing to have both TLSPeer.Listen and TLSPeer.ListenTLS

Same I guess with TLSPeer.Dial and TLSPeer.DialTLS

Reply:
ListenTLS and DialTLS are convenience wrappers when you don't want to manage the lifetime of the TLS Peer separately. For ListenTLS, the lifetime of the TLSPeer is managed internally by the listener. For DialTLS, the peer is used once and discarded.

Reply:
What's the use case for managing TLS Peer separately? Multiple listeners?

Reply:
That's one. Another is a peer that both listens and dials or does either repeatedly.

Reply:
Oops, just realized these two functions are static. That makes sense! 👍